fault_proof/bin/challenger.rs

```diff
@@ -0,0 +1,258 @@
+            if deadline < current_timestamp {
+                tracing::info!(
+                    "Game {:?} at index {:?} deadline {:?} has passed, not attempting to challenge",
+                    game_address,
+                    game_index,
+                    deadline
+                );
+                game_index += U256::from(1);
+                continue;
+            }
+
+            block_number = game.l2BlockNumber().call().await?.l2BlockNumber_;
+            tracing::info!(
+                "Checking if game {:?} at index {:?} for block {:?} is invalid",
+                game_address,
+                game_index,
+                block_number
+            );
+            let game_claim = game.rootClaim().call().await?.rootClaim_;
+
+            let output_root = self
+                .l2_provider
+                .compute_output_root_at_block(block_number)
+                .await?;
+
+            if output_root != game_claim {
+                tracing::info!(
+                    "Output root {:?} at block {:?} is not same as game claim {:?}",
+                    output_root,
+                    block_number,
+                    game_claim
+                );
+                break;
+            }
+
+            game_index += U256::from(1);
+        }
+
+        tracing::info!(
+            "Oldest challengable game {:?} at game index {:?} with l2 block number: {:?}",
+            game_address,
+            game_index,
+            block_number
+        );
+
+        Ok(Some(game_address))
+    }
+
+    /// Handles challenging of invalid games
+    async fn handle_game_challenging(&self) -> Result<()> {
+        let _span = tracing::info_span!("[[Challenging]]").entered();
+
+        if let Some(game_address) = self.get_oldest_challengable_game_address().await? {
+            tracing::info!("Attempting to challenge game {:?}", game_address);
+            self.challenge_game(game_address).await?;
+        }
+
+        Ok(())
+    }
+
+    /// Handles resolution of challenged games
+    async fn handle_game_resolution(&self) -> Result<()> {
+        // Only resolve games if the config is enabled
+        if !self.config.enable_game_resolution {
+            return Ok(());
+        }
+
+        let _span = tracing::info_span!("[[Resolving]]").entered();
+
+        self.factory
+            .resolve_games(
+                Mode::Challenger,
+                self.config.max_games_to_check_for_resolution,
+                self.l1_provider_with_wallet.clone(),
+                self.l2_provider.clone(),
+            )
+            .await
+    }
+
+    async fn run(&mut self) -> Result<()> {
+        tracing::info!("OP Succinct Challenger running...");
+        let mut interval = time::interval(Duration::from_secs(self.config.fetch_interval));
+
+        loop {
+            interval.tick().await;
+
+            if let Err(e) = self.handle_game_challenging().await {
```

Comment:
Each loop, we check the oldest challengeable game and challenge that. Eventually, we will challenge all games (as long as the rate at which games are being created is < than the interval.